akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorTimersSpec.scala

```diff
@@ -0,0 +1,132 @@
+            command match {
+              case "scheduled" =>
+                probe ! "scheduled"
+                Effect.none
+              case "cmd-0" =>
+                timers.startSingleTimer("key", "scheduled", Duration.Zero)
+                Effect.none
+              case _ =>
+                timers.startSingleTimer("key", "scheduled", Duration.Zero)
+                Effect.persist(command).thenRun(_ => probe ! command)
+            },
+          eventHandler = (state, evt) => state + evt)
+      }
+    }
+
+  def testTimerFromSetupBehavior(persistenceId: PersistenceId, probe: ActorRef[String]): Behavior[String] =
+    Behaviors.setup { _ =>
+      Behaviors.withTimers { timers =>
+        timers.startSingleTimer("key", "scheduled", Duration.Zero)
+
+        EventSourcedBehavior[String, String, String](
+          persistenceId,
+          emptyState = "",
+          commandHandler = (_, command) =>
+            command match {
+              case "scheduled" =>
+                probe ! "scheduled"
+                Effect.none
+              case _ =>
+                Effect.persist(command).thenRun(_ => probe ! command)
+            },
+          eventHandler = (state, evt) => state + evt)
+      }
+    }
+
+}
+
+class EventSourcedBehaviorTimersSpec
+    extends ScalaTestWithActorTestKit(EventSourcedBehaviorTimersSpec.config)
+    with WordSpecLike {
+
+  import EventSourcedBehaviorTimersSpec._
+
+  val pidCounter = new AtomicInteger(0)
+  private def nextPid(): PersistenceId = PersistenceId(s"c${pidCounter.incrementAndGet()})")
+
+  import akka.actor.typed.scaladsl.adapter._
+  // needed for the untyped event filter
+  private implicit val untypedSystem: akka.actor.ActorSystem = system.toUntyped
+
+  untypedSystem.eventStream.publish(Mute(EventFilter.warning(start = "No default snapshot store", occurrences = 1)))
+
+  "EventSourcedBehavior withTimers" must {
+
+    "be able to schedule message" in {
+      val probe = createTestProbe[String]()
+      val pid = nextPid()
+      val ref = spawn(testBehavior(pid, probe.ref))
+
+      ref ! "cmd-0"
+      probe.expectMessage("scheduled")
+    }
+
+    "not discard timer msg due to stashing" in {
+      val probe = createTestProbe[String]()
+      val pid = nextPid()
+      val ref = spawn(testBehavior(pid, probe.ref))
+
+      ref ! "cmd-1"
+      probe.expectMessage("cmd-1")
+      probe.expectMessage("scheduled")
+    }
+
+    "be able to schedule message from setup" in {
+      val probe = createTestProbe[String]()
+      val pid = nextPid()
+      val ref = spawn(testTimerFromSetupBehavior(pid, probe.ref))
+
+      probe.expectMessage("scheduled")
+
+      (1 to 20).foreach { n =>
+        ref ! s"cmd-$n"
+      }
+      probe.receiveMessages(20)
+
+      // start new instance that is likely to stash the timer message while replaying
+      spawn(testTimerFromSetupBehavior(pid, probe.ref))
+      probe.expectMessage("scheduled")
+    }
+
+  }
```

Comment:
I expect these last two are racy in whether they cover what they seem to cover, but I guess there isn't really any way around it, right? At least they aren't racy in a way that should fail now and then.

Reply:
yes, that's why I wrote "likely", and the outcome should be the same